packages/flutter_tools/lib/src/base/net.dart

```diff
@@ -12,31 +13,46 @@ const int kNetworkProblemExitCode = 50;
 
 /// Download a file from the given URL and return the bytes.
 Future<List<int>> fetchUrl(Uri url) async {
-  printTrace('Downloading $url.');
+  int attempts = 0;
+  int duration = 1;
+  while (true) {
+    attempts += 1;
+    final List<int> result = await _attempt(url);
+    if (result != null)
+      return result;
+    printStatus('Download failed -- attempting retry $attempts in $duration second${ duration == 1 ? "" : "s"}...');
+    await new Future<Null>.delayed(new Duration(seconds: duration));
+    if (duration < 64)
+      duration *= 2;
+  }
+}
 
-  final HttpClient httpClient = new HttpClient();
+Future<List<int>> _attempt(Uri url) async {
+  printTrace('Downloading: $url');
+  final HttpClient httpClient = context[HttpClient] ?? new HttpClient();
   final HttpClientRequest request = await httpClient.getUrl(url);
   final HttpClientResponse response = await request.close();
-
-  printTrace('Received response statusCode=${response.statusCode}');
   if (response.statusCode != 200) {
-    throwToolExit(
-      'Download failed: $url\n'
-          '  because (${response.statusCode}) ${response.reasonPhrase}',
-      exitCode: kNetworkProblemExitCode,
-    );
+    if (response.statusCode > 0 && response.statusCode < 500) {
+      throwToolExit(
+        'Download failed.\n'
+        'URL: $url\n'
+        'Error: ${response.statusCode} ${response.reasonPhrase}',
+        exitCode: kNetworkProblemExitCode,
+      );
+    }
+    // 5xx errors are server errors and we can try again
```

Comment:
What about http timeouts?  Does HttpClient take care of timing out, or is it up to us to wrap the future?  In either case, we probably want to handle a server not responding in the same way (exponential backoff retries).

Reply:
What's the difference between a server not responding, and trying to update Flutter on a really really bad connection?

Reply:
Good point